### PR TITLE
[PRO-272] I steven updates

### PR DESCRIPTION
**Pull request by jcrawley**
- adds support for infinite scrolling callback
- support for intermediate filtering state callback
- add support for virtual scrolling list (only keep the available lists on the page)

### CODE REVIEW

File: isteven-multi-select.js [hunk 1]
```diff
@@ -1,1111 +1,1377 @@
-/*
```
Comment (jcrawley) [inline isteven-multi-select.js]: Mostly a prettier reformat, I'll annotate changes

File: isteven-multi-select.js [hunk 2]
```diff
@@ -1,1111 +1,1377 @@
-/* 
+import 'angular-vs-repeat';
```
Comment (jcrawley) [inline isteven-multi-select.js:1]: vs-repeat is for scrolling render support

File: isteven-multi-select.js [hunk 3]
```diff
@@ -1,1111 +1,1377 @@
...
+          // callbacks
+          onClear: '&',
+          onClose: '&',
+          onSearchChange: '&',
+          onItemClick: '&',
+          onOpen: '&',
+          onReset: '&',
+          onSelectAll: '&',
+          onSelectNone: '&',
+          onScrollBottom: '&',
```
Comment (jcrawley) [inline isteven-multi-select.js]: `onScrollBottom`, `onBeginFiltering`, `onEndFiltering` are all new

File: isteven-multi-select.js [hunk 4]
```diff
@@ -1,1111 +1,1377 @@
...
-                // If available, then get select all, select none, and reset buttons
-                if ( $scope.helperStatus.all || $scope.helperStatus.none || $scope.helperStatus.reset ) {                                                       
-                    selectButtons = element.children().children().next().children().children()[ 0 ].getElementsByTagName( 'button' );                    
-                    // If available, then get the search box and the clear button
-                    if ( $scope.helperStatus.filter ) {                                            
-                        // Get helper - search and clear button. 
-                        inputField =    element.children().children().next().children().children().next()[ 0 ].getElementsByTagName( 'input' );                    
-                        clearButton =   element.children().children().next().children().children().next()[ 0 ].getElementsByTagName( 'button' );                        
-                    }
+            $scope.onBeginFiltering();
```
Comment (jcrawley) [inline isteven-multi-select.js]: added begin and end filtering call updates

File: isteven-multi-select.js [hunk 5]
```diff
@@ -1,1111 +1,1377 @@
...
-                        break;
-                    case 'CLEAR':
-                        $scope.tabIndex = $scope.tabIndex + 1;
-                        $scope.onClear();    
-                        break;
-                    case 'FILTER':                        
-                        $scope.tabIndex = helperItems.length - 1;
-                        break;
-                    default:                        
-                }
```
Comment (jcrawley) [inline isteven-multi-select.js]: added scroll listener for infinite scrolling logic

Comment (jcrawley) [inline isteven-multi-select.js]: vs-repeat div is used to encapsulate the ng-repeat. This is for scroll and rendering performance

File: isteven-multi-select.js (hunk 3 above)
Comment (nbaldwin) [reply isteven-multi-select.js]: Are we still using these now?

**Review by nbaldwin — APPROVED** [review]
🔥

File: isteven-multi-select.js [hunk 6]
```diff
@@ -1336,6 +1355,9 @@ angular
         '</div>' +
         '</div>' +
         '</div>' +
+        '</div>' +
```
Comment (creuter) [inline isteven-multi-select.js:1358]: Is there an extra closing div here?

File: package-lock.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (juniferd) [inline package-lock.json]: do we need this package-lock.json?

File: isteven-multi-select.js [hunk 8]
```diff
@@ -1336,6 +1355,9 @@ angular
...
+        '</div>' +
+        '</div>' +
         '</span>';
```
Comment (juniferd) [inline isteven-multi-select.js:1361]: i know this isn't your doing, but divs inside a span seems like a no-no

**Review by juniferd — APPROVED** [review]

File: package-lock.json (hunk 7 above)
Comment (jcrawley) [reply package-lock.json]: meh, I'll remove it

**Pull request merged**